README.md
# getfieldname README

This is the README for your extension "getfieldname". After writing up a brief description, we recommend including the following sections.

## Features

Describe specific features of your extension including screenshots of your extension in action. Image paths are relative to this README file.

For example if there is an image subfolder under your extension project workspace:

\!\[feature X\]\(images/feature-x.png\)

> Tip: Many popular extensions utilize animations. This is an excellent way to show off your extension! We recommend short, focused animations that are easy to follow.

## Requirements

If you have any requirements or dependencies, add a section describing those and how to install and configure them.

## Extension Settings

Include if your extension adds any VS Code settings through the `contributes.configuration` extension point.

For example:

This extension contributes the following settings:

* `myExtension.enable`: Enable/disable this extension.
* `myExtension.thing`: Set to `blah` to do something.

## Known Issues

Calling out known issues can help limit users opening duplicate issues against your extension.

## Release Notes

Users appreciate release notes as you update your extension.

### 1.0.0

Initial release of ...

### 1.0.1

Fixed issue #.

### 1.1.0

Added features X, Y, and Z.

---

## Following extension guidelines

Ensure that you've read through the extensions guidelines and follow the best practices for creating your extension.

* [Extension Guidelines](https://code.visualstudio.com/api/references/extension-guidelines)

## Working with Markdown

You can author your README using Visual Studio Code. Here are some useful editor keyboard shortcuts:

* Split the editor (`Cmd+\` on macOS or `Ctrl+\` on Windows and Linux).
* Toggle preview (`Shift+Cmd+V` on macOS or `Shift+Ctrl+V` on Windows and Linux).
* Press `Ctrl+Space` (Windows, Linux, macOS) to see a list of Markdown snippets.

## For more information

* [Visual Studio Code's Markdown Support](http://code.visualstudio.com/docs/languages/markdown)
* [Markdown Syntax Reference](https://help.github.com/articles/markdown-basics/)

**Enjoy!**


CHANGELOG.md
# Change Log

All notable changes to the "getfieldname" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/extension.ts
/**
 * todo
 * [done] gpt的 function call
 * [done] ‘/’触发SlashCommand，SlashCommand实现一个item类
 * [] 实现一个变量转换功能的item => ai 接口太慢了！
 * [] selection选区、空格问题 => 先解决上面问题再看这个问题！
 */

import * as vscode from "vscode";
import Suggestion, { SuggestionItemType } from "./suggestion/ index";
import VartranslatedEn from "./vartranslatedEn/index";

// 当前活跃的action
let isActiveSuggestionItem: string | undefined = undefined;

// ************************** 变量转换 vars **************************
// 中文words
let chineseWords: [number, number][] = [];
// currentActiveIndex
let currentActiveIndex = 0;
function nextActiveChineseWord() {
  if (currentActiveIndex >= chineseWords.length - 1) {
    currentActiveIndex = 0;
  } else {
    currentActiveIndex = currentActiveIndex + 1;
  }
  return chineseWords[currentActiveIndex];
}

// ************************** vscode utils **************************
/**
 * 获取单词的位置
 * @param character 参数表示一个字符的位置，它是一个数字类型（number）的值。
 * @param line 参数表示当前单词属于那一行，它是一个数字类型（number）的值
 * @returns 单词的位置
 */
function getWordPosition(character: number, line: number): vscode.Position {
  return new vscode.Position(line, character);
}

/**
 * 根据单词起始和结束位置设置选区
 * @param wordStartEndTuple 表示一个包含中文单词起始和结束位置的数组，它是一个元组类型（[number, number]）的值。
 * @param position 表示当前位置，它是一个vscode.Position类型的值。
 */
function setSelection(
  wordStartEndTuple: [number, number],
  position: vscode.Position
) {
  const startPosition = getWordPosition(wordStartEndTuple[0], position.line); // 设置起始位置
  const endPosition = getWordPosition(wordStartEndTuple[1], position.line); // 设置结束位置
  const selection = new vscode.Selection(startPosition, endPosition);
  const editor = vscode.window.activeTextEditor;
  editor!.selection = selection;
}

// ************************** vscode activate **************************

export async function activate(context: vscode.ExtensionContext) {
  /**
   * 由‘/’触发的action集合item
   */
  const getSuggestionItems = async (): Promise<SuggestionItemType[]> => {
    return [
      {
        label: "变量转换",
        kind: vscode.CompletionItemKind.Event,
        detail: '将中文变量译成英文变量',
        isHideInsertText: true,
        command: (document: vscode.TextDocument, position: vscode.Position) => {
          const editor = vscode.window.activeTextEditor;
          if (editor) {
            const position = editor.selection.active;
            const document = editor.document;
            const line = document.lineAt(position.line);
            const startPosition = new vscode.Position(
              position.line,
              line.range.end.character - 3
            ); // 设置起始位置
            const endPosition = new vscode.Position(
              position.line,
              line.range.end.character
            ); // 设置结束位置
            const selection = new vscode.Selection(startPosition, endPosition);
            editor.edit((editBuilder) => {
              editBuilder.delete(selection);
            });
          }
          
          const line = document.lineAt(position.line);
          const lineText = line.text;
          // 匹配中文单词的正则表达式
          const regex = /[\u4e00-\u9fa5]+/g;
          chineseWords = [];
          let match;
          while ((match = regex.exec(lineText)) !== null) {
            const startIndex = match.index;
            const endIndex = startIndex + match[0].length;
            const wordRange: [number, number] = [startIndex, endIndex];
            chineseWords.unshift(wordRange);
          }
          if (Array.isArray(chineseWords) && chineseWords.length > 0) {
            const firstchineseWord = chineseWords[0];
            setSelection(firstchineseWord, position);
            isActiveSuggestionItem = "变量转换";
          }
        },
      },
    ];
  };
  const suggestion = new Suggestion(getSuggestionItems);
  const disposables = suggestion.run();

  /**
   * 聚焦编辑器且有选中selection触发的tab键事件处理
   */
  let disposableTabListener = vscode.commands.registerTextEditorCommand(
    "extension.tabHasSelectionCommand",
    (textEditor) => {
      switch (isActiveSuggestionItem) {
        case "变量转换":
          const position = textEditor.selection.active;
          setSelection(nextActiveChineseWord(), position);
          break;

        default:
          vscode.commands.executeCommand("type", { text: "\t" });
          break;
      }
    }
  );

  /**
   * 聚焦编辑器且有选中selection触发的enter键事件处理
   */
  let disposableEnterListener = vscode.commands.registerTextEditorCommand(
    "extension.enterHasSelectionCommand",
    async (textEditor) => {
      const selection = textEditor.selection;
      const selectedText = textEditor.document.getText(selection);
      vscode.window.showInformationMessage(selectedText+ ' ' +'loading...');
      switch (isActiveSuggestionItem) {
        case "变量转换":
          isActiveSuggestionItem = undefined;
          const vartranslatedEn = new VartranslatedEn(selectedText);
          const vartranslatedEnDisposables = await vartranslatedEn.run();
          context.subscriptions.push(...vartranslatedEnDisposables);
          textEditor!.edit((editBuilder) => {
            editBuilder.delete(selection);
          });
          vscode.commands.executeCommand("editor.action.triggerSuggest");
          break;

        default:
          vscode.commands.executeCommand("editor.action.insertLineAfter");
          break;
      }
    }
  );

  context.subscriptions.push(
    ...disposables,
    disposableTabListener,
    disposableEnterListener
  );
}


src/suggestion/ index.ts
import * as vscode from "vscode";
import CustomCompletionItemProvider, {
  CommandItemProps,
} from "../customCompletionItem";

export type SuggestionItemType = CommandItemProps;

export default class Suggestion {
  constructor(
    private suggestionItems: (
      document: vscode.TextDocument,
      position: vscode.Position
    ) => Promise<Array<CommandItemProps>> | Array<CommandItemProps> = () => []
  ) {}
  public run() {
    const customCompletionItemProvider = new CustomCompletionItemProvider(
      this.suggestionItems,
      (lineText) => {
        return !lineText.endsWith("///");
      }
    );
    const disposable = vscode.languages.registerCompletionItemProvider(
      { scheme: "file" },
      customCompletionItemProvider,
      "/"
    );
    // 注册命令处理程序
    const commandDisposable = vscode.commands.registerCommand(
      "extension.itemSelected",
      async (command: () => void) => {
        await command();
      }
    );
    return [disposable, commandDisposable];
  }
}


src/transformVariable.ts
import * as vscode from 'vscode';
import fetch from 'node-fetch';

interface ContextType {
    language: 'typescript' | 'javascript',
    allText: string,
    relativePath?: string
}

export default async function transformVariable(
    document: vscode.TextDocument,
    position: vscode.Position,
    token: vscode.CancellationToken,
    context?: ContextType
): Promise<vscode.ProviderResult<vscode.Hover> | undefined> {
    if (!context) {
        context = {
            language: 'javascript',
            allText: ''
        };
    }
    const { language, allText, relativePath } = context;
    const shengmings = ['const', 'var', 'class', 'let', 'function'];
    const wordRange = document.getWordRangeAtPosition(position);
    if (!wordRange) {
        return;
    }
    const wordStart = wordRange!.start;
    const previousWordRange = document.getWordRangeAtPosition(wordStart.translate(0, -1));
    const previousWord = document.getText(previousWordRange);
    if (!shengmings.find(item => item === previousWord)) {
        return;
    }
    let config = vscode.workspace.getConfiguration('getfieldname');
    let value = config.get('openaikey');

    const hoveredWord = document.getText(wordRange);
    const response = await fetch('https://run.dingjunjie.com/v1/chat/completions', {
        method: "POST",
        headers: {
            // eslint-disable-next-line @typescript-eslint/naming-convention
            'Content-Type': "application/json",
            'authorization': "Bearer " + value,
        },
        body: JSON.stringify(
            {
                model: "gpt-3.5-turbo-16k-0613",
                temperature: 0,
                messages: [
                    {
                        role: "system",
                        content: `
                        用户是一名程序员，他(她)需要帮我处理一些事情，请严格按照她的要求执行，以下是代码的一些相关信息，你可以作为参考：
                        代码全文是：[${allText}],使用的语言是：${language}${relativePath?(',代码在项目所处路径是'+relativePath):''}。
                        `,
                    },
                    {
                        role: "user",
                        content: `我不太理解这个变量在这个代码中的意思：${hoveredWord}；它的意思是：`,
                    },
                ],
                stream: false,
            }
        ),
    });

    vscode.window.showInformationMessage((await response.json() as any).choices[0].message?.content);
    return await new Promise(resolve => {
        setTimeout(() => {
            const wordRange = document.getWordRangeAtPosition(position);
            const hoveredWord = document.getText(wordRange);
            resolve(new vscode.Hover(hoveredWord + '-' + previousWord));
        }, 0);
    });
}




src/vartranslatedEn/index.ts
import * as vscode from "vscode";
import fetch from "node-fetch";
import CustomCompletionItemProvider, {
  CommandItemProps,
} from "../customCompletionItem";

type ChoicesType = Array<{
  message: {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    function_call: {
      arguments: string;
    };
  };
}>;
type AiResurt = {
  choices: ChoicesType;
};

let config = vscode.workspace.getConfiguration("getfieldname");
let value = config.get("openaikey");

export default class VartranslatedEn {
  constructor(private selectedText: string) {}
  public async run() {
    try {
      if (!this?.selectedText) {
        vscode.window.showWarningMessage("未选中文本");
        return [];
      }
      const response = await fetch(
        "https://run.dingjunjie.com/v1/chat/completions",
        {
          method: "POST",
          headers: {
            // eslint-disable-next-line @typescript-eslint/naming-convention
            "Content-Type": "application/json",
            authorization:
              "Bearer " + value,
          },
          body: JSON.stringify({
            model: "gpt-3.5-turbo-16k-0613",
            messages: [
              {
                role: "user",
                content: `
                我在自学编程，但是我的母语不是英语，我需要将${this.selectedText} 字段改写成英文变量，你可以教我改吗？
                  `,
              },
            ],
            // eslint-disable-next-line @typescript-eslint/naming-convention
            function_call: {
              name: "get_var",
            },
            functions: [
              {
                name: "get_var",
                description: `            
                    根据当前中文变量，改写成英文变量，变量符合编程习惯
                          `,
                parameters: {
                  type: "object",
                  properties: {
                    varName1: {
                      type: "string",
                      description: "这里指改写后的英文变量，符合驼峰命名",
                    },
                    varName2: {
                      type: "string",
                      description: "这里同样是指改写后的英文变量，也符合驼峰命名，但它是备用变量",
                    },
                  },
                  required: ["varName1", "varName2"],
                },
              },
            ],
            stream: false,
          }),
        }
      );
      if (response.status !== 200) {
        const { error } = (await response.json()) as any;
        console.log(error, "error");
        vscode.window.showErrorMessage(error.message, response.statusText);
        return [];
      }
      const res = (await response.json()) as AiResurt;
      const responseMessage = res.choices[0].message;
      if (!responseMessage.function_call) {
        vscode.window.showErrorMessage("ai接口返回错误");
        return [];
      }
      const variablesStr = responseMessage.function_call.arguments;
      console.log("variablesStr===>", variablesStr);
      const variables: string[] = Object.values(JSON.parse(variablesStr));

      const getVartranslatedEnItem = (): CommandItemProps[] => {
        const vartranslatedEnItem = variables.map((item) => ({
          label: item,
          kind: vscode.CompletionItemKind.Event,
          detail: item,
          isHideInsertText: false,
        }));

        return vartranslatedEnItem;
      };
      const customCompletionItemProvider = new CustomCompletionItemProvider(
        getVartranslatedEnItem
      );
      const disposable = vscode.languages.registerCompletionItemProvider(
        { scheme: "file" },
        customCompletionItemProvider
      );
      return [disposable];
    } catch (error) {
      console.error(error);
      return [];
    }
  }
}
